goworker.go

```diff
@@ -87,12 +90,71 @@ func Init() error {
 // while they wait for an available connection. Expect this
 // API to change drastically.
 func GetConn() (*RedisConn, error) {
+	if workerSettings.RedisSettings.MasterName == "" {
+		return getConn()
+	} else {
+		return getConnSentinel()
+	}
+}
+
+func getConn() (*RedisConn, error) {
 	resource, err := pool.Get(ctx)
 
 	if err != nil {
 		return nil, err
 	}
 	return resource.(*RedisConn), nil
+
+}
+func getConnSentinel() (*RedisConn, error) {
+	deadConnection := errors.New("Dead connection")
+	slaveConnection := errors.New("Stale connection (to slave, not master)")
+	try := func() (*RedisConn, error) {
+		resource, err := pool.Get(ctx)
+
+		if err != nil {
+			return nil, err
+		}
+		conn := resource.(*RedisConn)
+
+		// close the connection and remove it from the pool so that new
+		// connections get created
+
+		// if the instance does not ping back
+		_, err = conn.Do("ping")
+		if err != nil {
+			conn.Close()
+			PutConn(nil)
+			return nil, deadConnection
+		}
+
+		// or if the instance is not a master anymore
+		role, err := redis.String(conn.Do("role"))
+		if err != nil {
+			PutConn(nil)
+			return nil, err
+		}
+		if role != "master" {
+			conn.Close()
+			PutConn(nil)
+			return nil, slaveConnection
+		}
+		return conn, nil
+	}
+
+	var conn *RedisConn
+	var err error
+	for i := 0; i < workerSettings.Connections+1; i++ {
+		if conn, err = try(); err == nil {
+			return conn, nil
+		} else if err != slaveConnection && err != deadConnection {
+			if err, ok := err.(net.Error); ok && !err.Timeout() {
```

Comment:
I think the condition might be wrong here. I believe that if we know it is a timeout, we want to try again (which is happening here). However, here, if we got an `err` that wasn't a `net.Error` it would continue retrying (i believe if we receive a non `net.Error` we'd want to return that immediately)